src/SourcePathsMappingUI/AskUserForFile.cpp

```diff
@@ -43,39 +46,61 @@ std::optional<QString> TryAskingTheUserAndReadSourceFile(QWidget* parent,
   // Ownership will be transferred to message_box
   message_box.setCheckBox(checkbox.release());
 
-  QString user_chosen_file;
+  std::optional<QString> maybe_file_contents;
   QObject::connect(pick_file_button, &QAbstractButton::clicked, parent, [&]() {
-    QDir previous_directory{
-        settings.value(kPreviousSourcePathsMappingDirectoryKey, QDir::currentPath()).toString()};
-    const QString file_name = QString::fromStdString(file_path.filename().string());
+    const bool infer_source_paths_mapping = message_box.checkBox()->isChecked();
+    ErrorMessageOr<std::optional<QString>> result =
+        ShowFileOpenDialogAndReadSourceFile(parent, file_path, infer_source_paths_mapping);
+    if (result.has_error()) {
+      QMessageBox::critical(parent, "Could not open source file",
+                            QString::fromStdString(result.error().message()));
+      return;
+    }
 
-    user_chosen_file = QFileDialog::getOpenFileName(
-        parent, QString{"Choose %1"}.arg(QString::fromStdString(file_path.string())),
-        previous_directory.filePath(file_name), file_name);
+    maybe_file_contents = std::move(result.value());
   });
 
   message_box.exec();
+
+  if (maybe_file_contents.has_value()) {
+    settings.setValue(kAutocreateMappingKey, message_box.checkBox()->isChecked());
+  }
+
+  return maybe_file_contents;
+}
+
+ErrorMessageOr<std::optional<QString>> ShowFileOpenDialogAndReadSourceFile(
```

Comment:
I find these "DoXAndY" methods a bit confusing and long. Wouldn't it be easier for the code flow to split them into two methods that each do one thing? Same for `TryAskingTheUserAndReadSourceFile`.

Reply:
I was thinking exactly the same. By the way, it's not always easy to do. I would prefer to have a method with all the logic together which call a bunch of small methods where each one only does one thing.

Reply:
I will try to split it up.

Reply:
Done. PTAL.

Reply:
LGTM!